internal/services/ingestor/buffered.go

```diff
@@ -0,0 +1,131 @@
+package ingestor
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hatchet-dev/hatchet/pkg/repository"
+	"github.com/hatchet-dev/hatchet/pkg/repository/prisma/dbsqlc"
+)
+
+type IngestBuf struct {
+	maxCapacity        int           // max number of events to hold in buffer before we flush
+	flushPeriod        time.Duration // max time to hold events in buffer before we flush
+	eventOpsChan       chan *eventBuffWrapper
+	lastFlush          time.Time
+	internalArr        []*eventBuffWrapper
+	sizeOfData         int // size of data in buffer
+	maxDataSizeInQueue int // max number of bytes to hold in buffer before we flush
+	BulkCreateFunc     func(ctx context.Context, opts *repository.BulkCreateEventSharedTenantOpts) (*repository.BulkCreateEventResult, error)
+}
+
+func NewIngestBuffer(maxCapacity int, flushPeriod time.Duration, maxDataSizeInQueue int, bulkCreateFunc func(ctx context.Context, opts *repository.BulkCreateEventSharedTenantOpts) (*repository.BulkCreateEventResult, error)) *IngestBuf {
+	return &IngestBuf{
+		maxCapacity:        maxCapacity,
+		flushPeriod:        flushPeriod,
+		eventOpsChan:       make(chan *eventBuffWrapper, maxCapacity*2),
+		lastFlush:          time.Now(),
+		internalArr:        make([]*eventBuffWrapper, 0),
+		sizeOfData:         0,
+		maxDataSizeInQueue: maxDataSizeInQueue,
+		BulkCreateFunc:     bulkCreateFunc,
+	}
+}
+
+type eventBuffWrapper struct {
+	eventOps *repository.CreateEventOpts
+	doneChan chan *flushResponse
+}
+
+func (b *IngestBuf) buffEventWorker(ctx context.Context) {
+	for {
+		select {
+		case e := <-b.eventOpsChan:
+			b.internalArr = append(b.internalArr, e)
+			b.sizeOfData += len(e.eventOps.Data)
+			b.sizeOfData += len(e.eventOps.AdditionalMetadata)
+
+			if len(b.internalArr) >= b.maxCapacity {
+				b.flush()
```

Comment:
It seems the `flush` is blocking and waits for writes to succeed before exiting. Does it make sense to make `flush` non-blocking?

In that case we will need to ensure that `internalArr` is rw concurrency-safe, or copy the contents of `internalArr` over to the flush method.

Reply:
good idea, was thinking that we wanted some back pressure to control writes to the DB, but in the case where the max buffer size is one (or low) we want to write straight away - and what we are replacing is a write straight away - so it should not be too much for the DB to handle.